src/range.rs

```diff
@@ -0,0 +1,355 @@
+        impl TryFrom<RangeInclusive<$lte>> for TextRange {
+            type Error = TryFromIntError;
+            fn try_from(value: RangeInclusive<$lte>) -> Result<TextRange, Self::Error> {
+                let (start, end) = value.into_inner();
+                let end: TextSize = end.into();
+                // This is the only way to get a TryFromIntError currently.
+                let end = end.checked_add(1).ok_or_else(|| u8::try_from(-1).unwrap_err())?;
+                Ok(TextRange(start.into(), end))
+            }
+        }
+        impl From<RangeTo<$lte>> for TextRange {
+            fn from(value: RangeTo<$lte>) -> TextRange {
+                TextRange(0.into(), value.end.into())
+            }
+        }
+        impl TryFrom<RangeToInclusive<$lte>> for TextRange {
+            type Error = TryFromIntError;
+            fn try_from(value: RangeToInclusive<$lte>) -> Result<TextRange, Self::Error> {
+                let start: TextSize = 0.into();
+                let end: TextSize = value.end.into();
+                TextRange::try_from(start..=end)
+            }
+        }
+    };
+    (TryFrom<$gt:ident> for TextRange) => {
+        impl TryFrom<Range<$gt>> for TextRange {
+            type Error = <$gt as TryInto<u32>>::Error;
+            fn try_from(value: Range<$gt>) -> Result<TextRange, Self::Error> {
+                Ok(TextRange(value.start.try_into()?, value.end.try_into()?))
+            }
+        }
+        impl TryFrom<RangeInclusive<$gt>> for TextRange {
+            type Error = TryFromIntError;
+            fn try_from(value: RangeInclusive<$gt>) -> Result<TextRange, Self::Error> {
+                let (start, end) = value.into_inner();
+                let end: TextSize = end.try_into()?;
+                // This is the only way to get a TryFromIntError currently.
+                let end = end.checked_add(1).ok_or_else(|| u8::try_from(-1).unwrap_err())?;
+                Ok(TextRange(start.try_into()?, end))
+            }
+        }
+        impl TryFrom<RangeTo<$gt>> for TextRange {
+            type Error = TryFromIntError;
+            fn try_from(value: RangeTo<$gt>) -> Result<TextRange, Self::Error> {
+                Ok(TextRange(0.into(), value.end.try_into()?))
+            }
+        }
+        impl TryFrom<RangeToInclusive<$gt>> for TextRange {
+            type Error = TryFromIntError;
+            fn try_from(value: RangeToInclusive<$gt>) -> Result<TextRange, Self::Error> {
+                let start: TextSize = 0.into();
+                let end: TextSize = value.end.try_into()?;
+                TextRange::try_from(start..=end)
+            }
+        }
+    };
+    {
+        lt TextSize [$($lt:ident)*]
+        eq TextSize [$($eq:ident)*]
+        gt TextSize [$($gt:ident)*]
+        varries     [$($var:ident)*]
+    } => {
+        $(
+            // Not `From` yet because of integer type fallback. We want e.g.
+            // `TextRange::from(0)` and `range + 1` to work, and more `From`
```

Comment:
> `range + 1` to work

I think we explicitly want this **not** to work: it's only meaningful to shift ranges by several utf8_lengths of the string they originated from. `range + 1`  might, or might not be valid depending on whether the original string was ascii, and that seems like fragile API

Reply:
>or it to produce a correct range, the offset has to be a valid utf8 boundary for both the start and the end index. (This is the currently implemented behavior.) 

This is if you want to shift a range withing a string, which is indeed a nonsense operation. However, +/- are useful when you want to shift range **with** the string, when going from relative to absolute coordinates. Ie, if you have a `range: TextRange` inside a node, then `range + node.text_range().start()` will give you an range in file.